rpcserver.go

```diff
@@ -3324,18 +3324,33 @@ func handleSendRawTransaction(s *rpcServer, cmd interface{}, closeChan <-chan st
 		// When the error is a rule error, it means the transaction was
 		// simply rejected as opposed to something actually going wrong,
 		// so log it as such.  Otherwise, something really did go wrong,
-		// so log it as an actual error.  In both cases, a JSON-RPC
-		// error is returned to the client with the deserialization
-		// error code (to match bitcoind behavior).
-		if _, ok := err.(mempool.RuleError); ok {
+		// so log it as an actual error.
+		ruleErr, ok := err.(mempool.RuleError)
+		if ok {
 			rpcsLog.Debugf("Rejected transaction %v: %v", tx.Hash(),
 				err)
 		} else {
 			rpcsLog.Errorf("Failed to process transaction %v: %v",
 				tx.Hash(), err)
 		}
+
+		// We'll then map the rule error to the appropriate RPC error,
+		// matching bitcoind's behavior.
+		code := btcjson.ErrRPCTxError
+		if txRuleErr, ok := ruleErr.Err.(mempool.TxRuleError); ok {
+			errDesc := txRuleErr.Description
+			switch {
+			case strings.Contains(errDesc, "orphan transaction"):
```

Comment:
can we add error codes to `RuleError` instead of doing string matching here?

Reply:
No unless we return more structured errors from `TxPool.MaybeAcceptTransaction`. The current errors use the `wire.RejectCode`s, which don't provide enough context as they can be used for different scenarios. For example, `wire.RejectDuplicate` is used for transactions that have already confirmed, transactions that are already in the mempool, and transactions double spending others in the mempool.

Reply:
Perhaps we should go ahead and output better errors? Once again we find ourselves string matching here without any existing `btcd` level integration tests for this particular scenario for this call.

Reply:
Agreed. Is it worth deferring until we overhaul `btcwallet` with concrete errors? Can also use the newest error additions that will come as part of `go 1.13`, which will make things a bit nicer.